### PR TITLE
Additional Istio resources into the reconciliation

### PR DESCRIPTION
**Pull request by videlov**
**Description**

Changes proposed in this pull request:

- Reconciles [additional Istio resources](https://github.com/kyma-project/kyma/tree/main/resources/istio-resources) like Kyma default Gateway, Grafana Dashboards, Virtual Service for Availability endpoint, PeerAuthentication with STRICT mTLS, etc.

**Related issue(s)**

https://github.com/kyma-project/istio/issues/334

### CODE REVIEW

File: internal/reconciliations/istio_resources/configmap_control_plane.yaml [hunk 1]
```diff
@@ -0,0 +1,1763 @@
...
+            "datasource": "-- Grafana --",
+            "enable": true,
+            "hide": true,
+            "iconColor": "rgba(0, 211, 255, 1)",
+            "name": "Annotations & Alerts",
+            "type": "dashboard"
+          }
+        ]
+      },
+      "description": "Istio Control Plane Dashboard version 1.17.1",
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_control_plane.yaml:25]: Honestly, we might just want to remove the version from the dashboards
```suggestion
      "description": "Istio Control Plane Dashboard version 1.18.2",
```

File: internal/reconciliations/istio_resources/configmap_mesh.yaml [hunk 2]
```diff
@@ -0,0 +1,1758 @@
...
+            "datasource": "-- Grafana --",
+            "enable": true,
+            "hide": true,
+            "iconColor": "rgba(0, 211, 255, 1)",
+            "name": "Annotations & Alerts",
+            "type": "dashboard"
+          }
+        ]
+      },
+      "description": "Istio Mesh Dashboard version 1.17.1",
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_mesh.yaml:25]:
```suggestion
      "description": "Istio Mesh Dashboard version 1.18.2",
```

File: internal/reconciliations/istio_resources/configmap_performance.yaml [hunk 3]
```diff
@@ -0,0 +1,1328 @@
...
+            "datasource": "-- Grafana --",
+            "enable": true,
+            "hide": true,
+            "iconColor": "rgba(0, 211, 255, 1)",
+            "name": "Annotations & Alerts",
+            "type": "dashboard"
+          }
+        ]
+      },
+      "description": "Istio Performance Dashboard version 1.17.1",
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_performance.yaml:25]:
```suggestion
      "description": "Istio Performance Dashboard version 1.18.2",
```

File: internal/reconciliations/istio_resources/configmap_service.yaml [hunk 4]
```diff
@@ -0,0 +1,3004 @@
...
+            "datasource": "-- Grafana --",
+            "enable": true,
+            "hide": true,
+            "iconColor": "rgba(0, 211, 255, 1)",
+            "name": "Annotations & Alerts",
+            "type": "dashboard"
+          }
+        ]
+      },
+      "description": "Istio Service Dashboard version 1.17.1",
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_service.yaml:25]:
```suggestion
      "description": "Istio Service Dashboard version 1.18.2",
```

File: internal/reconciliations/istio_resources/configmap_workload.yaml [hunk 5]
```diff
@@ -0,0 +1,2686 @@
...
+            "datasource": "-- Grafana --",
+            "enable": true,
+            "hide": true,
+            "iconColor": "rgba(0, 211, 255, 1)",
+            "name": "Annotations & Alerts",
+            "type": "dashboard"
+          }
+        ]
+      },
+      "description": "Istio Workload Dashboard version 1.17.1",
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_workload.yaml:25]:
```suggestion
      "description": "Istio Workload Dashboard version 1.18.2",
```

File: internal/reconciliations/istio_resources/gateway_kyma.yaml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      number: 443
+      name: https
+      protocol: HTTPS
+    tls:
+      mode: SIMPLE
+      credentialName: kyma-gateway-certs
+      cipherSuites:
+      - ECDHE-RSA-CHACHA20-POLY1305
+      - ECDHE-RSA-AES256-GCM-SHA384
+      - ECDHE-RSA-AES128-GCM-SHA256
```
Comment (barchw) [inline internal/reconciliations/istio_resources/gateway_kyma.yaml:21]: We should be able to remove those
```suggestion
```

**Review by barchw — CHANGES_REQUESTED** [review]

File: internal/reconciliations/istio_resources/configmap_control_plane.yaml (hunk 1 above)
Comment (videlov) [reply internal/reconciliations/istio_resources/configmap_control_plane.yaml:25]: Yes, agree, lets remove this version.

File: internal/reconciliations/istio_resources/gateway_kyma.yaml (hunk 6 above)
Comment (barchw) [reply internal/reconciliations/istio_resources/gateway_kyma.yaml:21]: Yep, those are default see: https://github.com/orgs/kyma-project/projects/27/views/1?pane=issue&itemId=30035511

File: docs/contributor/04-40-istio-resources.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
+# Kyma Istio Additional Resources
+
+## Overview
+
+The Istio additional resources includes Kyma configuration of Istio and consists of:
```
Comment (nataliasitko) [inline docs/contributor/04-40-istio-resources.md:5]:
```suggestion
The additional Istio resources include the Kyma configuration of Istio. They consist of:
```

File: docs/contributor/04-40-istio-resources.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+- Istio monitoring configuration details providing Grafana dashboards specification
```
Comment (nataliasitko) [inline docs/contributor/04-40-istio-resources.md:7]:
```suggestion

- Configuration details for Istio monitoring containing specifications for Grafana dashboards
```

File: docs/contributor/04-40-istio-resources.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+- Istio Ingress Gateway configuring incoming traffic to Kyma
+- Mutual TLS (mTLS) configuration enabling mTLS cluster-wide in the STRICT mode
+- Istio [VirtualService](https://istio.io/docs/reference/config/networking/virtual-service/) informing whether Istio is up and running
```
Comment (nataliasitko) [inline docs/contributor/04-40-istio-resources.md:10]:
```suggestion
- Information about Istio [VirtualService](https://istio.io/docs/reference/config/networking/virtual-service/), which indicates whether Istio is operational.
```

File: docs/release-notes/1.0.1.md [hunk 10]
```diff
@@ -0,0 +1,5 @@
+
+## New features
+
+- Apply additional Kyma resources and configuration like Kyma Gateway, configure service-mesh traffic to be mTLS only with PeerAuthentication, install Istio Grafana dashboards and expose Istio availability endpoint with a Virtual Service [#334](https://github.com/kyma-project/istio/issues/334)
```
Comment (nataliasitko) [inline docs/release-notes/1.0.1.md:4]:
```suggestion
- Apply additional Kyma resources and configuration like Kyma Gateway, configure service-mesh traffic to only allow mTLS using PeerAuthentication, install Istio Grafana dashboards, and expose Istio availability endpoint with a Virtual Service [#334](https://github.com/kyma-project/istio/issues/334)
```

File: tests/integration/hooks.go [hunk 11]
```diff
@@ -40,8 +43,88 @@ var testObjectsTearDown = func(ctx context.Context, sc *godog.Scenario, _ error)
 	return ctx, nil
 }
 
-var istioCrTearDown = func(ctx context.Context, sc *godog.Scenario, _ error) (context.Context, error) {
+var additionalResourcesTearDown = func(ctx context.Context, sc *godog.Scenario, _ error) (context.Context, error) {
```
Comment (barchw) [inline tests/integration/hooks.go:46]: I think we should add them here: https://github.com/kyma-project/istio/tree/main/manifests instead

Comment (barchw) [reply tests/integration/hooks.go:46]: Or maybe putting an OwnerReference on the ConfigMaps would also work, so k8s will remove them for us

Comment (videlov) [reply tests/integration/hooks.go:46]: yes, good idea, lets try that.

File: internal/reconciliations/istio_resources/configmap_control_plane.go [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+	result, err := controllerutil.CreateOrUpdate(ctx, k8sClient, &resource, func() error {
+		resource.Object["spec"] = spec
+		resource.SetOwnerReferences([]metav1.OwnerReference{owner})
+		return nil
+	})
+	if err != nil {
+		return controllerutil.OperationResultNone, err
+	}
+
+	var daFound bool
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_control_plane.go:43]:
```suggestion
	var daFound bool = false
```

File: internal/reconciliations/istio_resources/configmap_control_plane.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+		if err != nil {
+			return controllerutil.OperationResultNone, err
+		}
+	}
+
+	return result, nil
+}
+
+func (ConfigMapControlPlane) Name() string {
+	return "config map control plane"
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_control_plane.go:58]:
```suggestion
	return "ConfigMap/istio-control-plane-grafana-dashboard"
```

File: internal/reconciliations/istio_resources/configmap_mesh.go [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+	result, err := controllerutil.CreateOrUpdate(ctx, k8sClient, &resource, func() error {
+		resource.Object["spec"] = spec
+		resource.SetOwnerReferences([]metav1.OwnerReference{owner})
+		return nil
+	})
+	if err != nil {
+		return controllerutil.OperationResultNone, err
+	}
+
+	var daFound bool
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_mesh.go:43]:
```suggestion
	var daFound bool = false
```

File: internal/reconciliations/istio_resources/configmap_mesh.go [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+		if err != nil {
+			return controllerutil.OperationResultNone, err
+		}
+	}
+
+	return result, nil
+}
+
+func (ConfigMapMesh) Name() string {
+	return "config map mesh"
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_mesh.go:58]:
```suggestion
	return "ConfigMap/istio-mesh-grafana-dashboard"
```

File: internal/reconciliations/istio_resources/configmap_performance.go [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+		if err != nil {
+			return controllerutil.OperationResultNone, err
+		}
+	}
+
+	return result, nil
+}
+
+func (ConfigMapPerformance) Name() string {
+	return "config map performance"
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_performance.go:58]:
```suggestion
	return "ConfigMap/istio-performance-grafana-dashboard"
```

File: internal/reconciliations/istio_resources/configmap_service.go [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+	result, err := controllerutil.CreateOrUpdate(ctx, k8sClient, &resource, func() error {
+		resource.Object["spec"] = spec
+		resource.SetOwnerReferences([]metav1.OwnerReference{owner})
+		return nil
+	})
+	if err != nil {
+		return controllerutil.OperationResultNone, err
+	}
+
+	var daFound bool
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_service.go:43]:
```suggestion
	var daFound bool = false
```

File: internal/reconciliations/istio_resources/configmap_service.go [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+		if err != nil {
+			return controllerutil.OperationResultNone, err
+		}
+	}
+
+	return result, nil
+}
+
+func (ConfigMapService) Name() string {
+	return "config map service"
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_service.go:58]:
```suggestion
	return "ConfigMap/istio-service-grafana-dashboard"
```

File: internal/reconciliations/istio_resources/configmap_workload.go [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+	result, err := controllerutil.CreateOrUpdate(ctx, k8sClient, &resource, func() error {
+		resource.Object["spec"] = spec
+		resource.SetOwnerReferences([]metav1.OwnerReference{owner})
+		return nil
+	})
+	if err != nil {
+		return controllerutil.OperationResultNone, err
+	}
+
+	var daFound bool
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_workload.go:43]:
```suggestion
	var daFound bool = false
```

File: internal/reconciliations/istio_resources/configmap_workload.go [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+		if err != nil {
+			return controllerutil.OperationResultNone, err
+		}
+	}
+
+	return result, nil
+}
+
+func (ConfigMapWorkload) Name() string {
+	return "config map workload"
```
Comment (barchw) [inline internal/reconciliations/istio_resources/configmap_workload.go:58]:
```suggestion
	return "ConfigMap/istio-workload-grafana-dashboard"
```

File: internal/reconciliations/istio_resources/envoy_filter_allow_partial_referer.go [hunk 21]
```diff
@@ -63,17 +66,26 @@ func (e EnvoyFilterAllowPartialReferer) apply(ctx context.Context, k8sClient cli
 	if err != nil {
 		return controllerutil.OperationResultNone, err
 	}
-	var ok bool
-	if envoyFilter.GetAnnotations() != nil {
-		_, ok = envoyFilter.GetAnnotations()[EnvoyFilterAnnotation]
+
+	var efaFound, daFound bool
```
Comment (barchw) [inline internal/reconciliations/istio_resources/envoy_filter_allow_partial_referer.go:70]:
```suggestion
	var efaFound, daFound bool = false, false
```

File: internal/reconciliations/istio_resources/gateway_kyma.go [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+		if err != nil {
+			return controllerutil.OperationResultNone, err
+		}
+	}
+
+	return result, nil
+}
+
+func (GatewayKyma) Name() string {
+	return "gateway kyma"
```
Comment (barchw) [inline internal/reconciliations/istio_resources/gateway_kyma.go:70]:
```suggestion
	return "Gateway/kyma-gateway"
```

File: internal/reconciliations/istio_resources/peer_authentication_mtls.go [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+	spec := resource.Object["spec"]
+	result, err := controllerutil.CreateOrUpdate(ctx, k8sClient, &resource, func() error {
+		resource.Object["spec"] = spec
+		return nil
+	})
+	if err != nil {
+		return controllerutil.OperationResultNone, err
+	}
+
+	var daFound bool
```
Comment (barchw) [inline internal/reconciliations/istio_resources/peer_authentication_mtls.go:42]:
```suggestion
	var daFound bool = false
```

File: internal/reconciliations/istio_resources/peer_authentication_mtls.go [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+		if err != nil {
+			return controllerutil.OperationResultNone, err
+		}
+	}
+
+	return result, nil
+}
+
+func (PeerAuthenticationMtls) Name() string {
+	return "peer authentication mtls"
```
Comment (barchw) [inline internal/reconciliations/istio_resources/peer_authentication_mtls.go:57]:
```suggestion
	return "PeerAuthentication/default"
```

File: tests/integration/scenario.go [hunk 25]
```diff
@@ -40,6 +40,7 @@ func initScenario(ctx *godog.ScenarioContext) {
 	ctx.Step(`^Virtual service "([^"]*)" exposing service "([^"]*)" by gateway "([^"]*)" is configured in namespace "([^"]*)"$`, steps.CreateVirtualService)
 	ctx.Step(`^Request with header X-Forwarded-For with value "([^"]*)" sent to httpbin should return X-Envoy-External-Address with value "([^"]*)"$`, steps.ValidateHeader)
 	ctx.Step(`^Request with header "([^"]*)" with value "([^"]*)" sent to httpbin should return "([^"]*)" with value "([^"]*)"$`, steps.ValidateHeader)
+	ctx.Step(`^"([^"]*)" "([^"]*)" in namespace "([^"]*)" is applied`, steps.ResourceIsApplied)
```
Comment (barchw) [inline tests/integration/scenario.go:43]: Applied might suggest that the resource is applied as part of the test run
```suggestion
	ctx.Step(`^"([^"]*)" "([^"]*)" in namespace "([^"]*)" is present`, steps.ResourceIsApplied)
```

**Review by barchw — CHANGES_REQUESTED** [review]
There's a lot of code duplication (especially for ConfigMap applies). Maybe we could extract some parts that are the same?

File: internal/reconciliations/istio_resources/configmap_control_plane.go (hunk 12 above)
Comment (videlov) [reply internal/reconciliations/istio_resources/configmap_control_plane.go:43]: default bool value in go is false?

**Comment by videlov** [comment]
> There's a lot of code duplication (especially for ConfigMap applies). Maybe we could extract some parts that are the same?

agree, but I won't able to refactor this today and someone else would have to take over, would you ?

File: internal/reconciliations/istio_resources/configmap_control_plane.go (hunk 12 above)
Comment (videlov) [reply internal/reconciliations/istio_resources/configmap_control_plane.go:43]: preference thing, let be happy 😊

File: tests/integration/features/istio/production/main-suite/istio_resources.feature [hunk 26]
```diff
@@ -4,14 +4,13 @@ Feature: Istio resources configuration
...
     And Istio CRD is installed
-    And "Deployment" "istio-controller-manager" in namespace "kyma-system" is ready
 
   Scenario: Additional Istio resources are present
     Given Istio CR "istio-sample" is applied in namespace "kyma-system"
     When Istio CR "istio-sample" in namespace "kyma-system" has status "Ready"
     Then "Gateway" "kyma-gateway" in namespace "kyma-system" is present
     And "EnvoyFilter" "kyma-referer" in namespace "istio-system" is present
-    And "PeerAuthentication" "defaul" in namespace "istio-system" is present
+    And "PeerAuthentication" "default" in namespace "istio-system" is present
```
Comment (videlov) [inline tests/integration/features/istio/production/main-suite/istio_resources.feature:14]: good catch 👍

**Review by barchw — APPROVED** [review]

**Review by nataliasitko — APPROVED** [review]

**Pull request merged**